medcat/stats/kfold.py

```diff
@@ -299,52 +300,72 @@ def get_per_fold_metrics(cat: CATLike, folds: List[MedCATTrainerExport],
     return metrics
 
 
-def _update_all_weighted_average(joined: List[Dict[str, Tuple[int, float]]],
-                single: List[Dict[str, float]], cui2count: Dict[str, int]) -> None:
-    if len(joined) != len(single):
-        raise ValueError(f"Incompatible lists. Joined {len(joined)} and single {len(single)}")
-    for j, s in zip(joined, single):
-        _update_one_weighted_average(j, s, cui2count)
+def _merge_examples(all_examples: Dict, cur_examples: Dict) -> None:
+    for ex_type, ex_dict in cur_examples.items():
+        if ex_type not in all_examples:
+            all_examples[ex_type] = {}
+        per_type_examples = all_examples[ex_type]
+        for ex_cui, cui_examples_list in ex_dict.items():
+            if ex_cui not in per_type_examples:
+                per_type_examples[ex_cui] = []
+            per_type_examples[ex_cui].extend(cui_examples_list)
+
+
+# helper types
+IntValuedMetric = Union[
+    Dict[str, int],
+    Dict[str, Tuple[int, float]]
+]
+FloatValuedMetric = Union[
+    Dict[str, float],
+    Dict[str, Tuple[float, float]]
+]
+
 
+class PerCUIMetrics(BaseModel):
+    weights: List[Union[int, float]] = []
+    vals: List[Union[int, float]] = []
 
-def _update_one_weighted_average(joined: Dict[str, Tuple[int, float]],
-                one: Dict[str, float],
-                cui2count: Dict[str, int]) -> None:
-    for k in one:
-        if k not in joined:
-            joined[k] = (0, 0)
-        prev_w, prev_val = joined[k]
-        new_w, new_val = cui2count[k], one[k]
-        total_w = prev_w + new_w
-        total_val = (prev_w * prev_val + new_w * new_val) / total_w
-        joined[k] = (total_w, total_val)
+    def add(self, val, weight: int = 1):
+        self.weights.append(weight)
+        self.vals.append(val)
 
+    def get_mean(self):
+        return sum(w * v for w, v in zip(self.weights, self.vals)) / sum(self.weights)
 
-def _update_all_add(joined: List[Dict[str, int]], single: List[Dict[str, int]]) -> None:
+    def get_std(self):
```

Comment:
Is this needed given this: https://docs.python.org/3.11/library/statistics.html#statistics.pstdev
you can specify `mu=get_mean()`.

Also in 3.11, `statistics.fmean` supports a `weigthing`

Reply:
The current implementation expects to work for the weighted case (it's just that in some cases the weights are all equal (to 1)).
As far as I can tell, the other options do not (at least currently) support weighted means or standard deviations.

Plus, we'd need something that would support 3.8 through to 3.11.